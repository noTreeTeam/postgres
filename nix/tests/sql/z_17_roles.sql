-- version-specific roles and attributes
select
  rolname,
  rolcreaterole  ,
  rolcanlogin    ,
  rolsuper       ,
  rolinherit     ,
  rolcreatedb    ,
  rolreplication ,
  rolconnlimit   ,
  rolbypassrls   ,
  rolvaliduntil
from pg_roles r
where rolname in ('pg_create_subscription', 'pg_maintain', 'pg_use_reserved_connections')
order by rolname;

select
  rolname,
  rolconfig
from pg_roles r
where rolname in ('pg_create_subscription', 'pg_maintain', 'pg_use_reserved_connections')
order by rolname;

-- version-specific role memberships
select
    r.rolname as member,
    g.rolname as "member_of (can become)",
    m.admin_option
from
    pg_roles r
join
    pg_auth_members m on r.oid = m.member
left join
    pg_roles g on m.roleid = g.oid
order by
    r.rolname, g.rolname;

-- Check version-specific privileges of the roles on the schemas
select schema_name, privilege_type, grantee, default_for
from (
    -- ALTER DEFAULT privileges on schemas
    select
        n.nspname as schema_name,
        a.privilege_type,
        r.rolname as grantee,
        d.defaclrole::regrole as default_for,
        case when n.nspname = 'public' then 0 else 1 end as schema_order
    from
        pg_default_acl d
    join
        pg_namespace n on d.defaclnamespace = n.oid
    cross join lateral aclexplode(d.defaclacl) as a
    join
        pg_roles r on a.grantee = r.oid
    where
        a.privilege_type = 'MAINTAIN'
) sub
order by schema_order, schema_name, privilege_type, grantee, default_for;

-- version specific role memberships
select
    r.rolname as member,
    g.rolname as "member_of (can become)",
    m.admin_option
from
    pg_roles r
left join
    pg_auth_members m on r.oid = m.member
left join
    pg_roles g on m.roleid = g.oid
where r.rolname not in ('pg_create_subscription', 'pg_maintain', 'pg_use_reserved_connections')
and g.rolname not in ('pg_create_subscription', 'pg_maintain', 'pg_use_reserved_connections')
order by
    r.rolname, g.rolname;
