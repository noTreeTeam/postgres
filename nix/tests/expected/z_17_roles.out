-- version-specific roles and attributes
select
  rolname,
  rolcreaterole  ,
  rolcanlogin    ,
  rolsuper       ,
  rolinherit     ,
  rolcreatedb    ,
  rolreplication ,
  rolconnlimit   ,
  rolbypassrls   ,
  rolvaliduntil
from pg_roles r
where rolname in ('pg_create_subscription', 'pg_maintain', 'pg_use_reserved_connections')
order by rolname;
           rolname           | rolcreaterole | rolcanlogin | rolsuper | rolinherit | rolcreatedb | rolreplication | rolconnlimit | rolbypassrls | rolvaliduntil 
-----------------------------+---------------+-------------+----------+------------+-------------+----------------+--------------+--------------+---------------
 pg_create_subscription      | f             | f           | f        | t          | f           | f              |           -1 | f            | 
 pg_maintain                 | f             | f           | f        | t          | f           | f              |           -1 | f            | 
 pg_use_reserved_connections | f             | f           | f        | t          | f           | f              |           -1 | f            | 
(3 rows)

select
  rolname,
  rolconfig
from pg_roles r
where rolname in ('pg_create_subscription', 'pg_maintain', 'pg_use_reserved_connections')
order by rolname;
           rolname           | rolconfig 
-----------------------------+-----------
 pg_create_subscription      | 
 pg_maintain                 | 
 pg_use_reserved_connections | 
(3 rows)

-- version-specific role memberships
select
    r.rolname as member,
    g.rolname as "member_of (can become)",
    m.admin_option
from
    pg_roles r
left join
    pg_auth_members m on r.oid = m.member
left join
    pg_roles g on m.roleid = g.oid
where r.rolname in ('pg_create_subscription', 'pg_maintain', 'pg_use_reserved_connections')
or g.rolname in ('pg_create_subscription', 'pg_maintain', 'pg_use_reserved_connections')
order by
    r.rolname, g.rolname;
           member            | member_of (can become) | admin_option 
-----------------------------+------------------------+--------------
 pg_create_subscription      |                        | 
 pg_maintain                 |                        | 
 pg_use_reserved_connections |                        | 
 postgres                    | pg_create_subscription | f
(4 rows)

-- Check version-specific privileges of the roles on the schemas
select schema_name, privilege_type, grantee, default_for
from (
    -- ALTER DEFAULT privileges on schemas
    select
        n.nspname as schema_name,
        a.privilege_type,
        r.rolname as grantee,
        d.defaclrole::regrole as default_for,
        case when n.nspname = 'public' then 0 else 1 end as schema_order
    from
        pg_default_acl d
    join
        pg_namespace n on d.defaclnamespace = n.oid
    cross join lateral aclexplode(d.defaclacl) as a
    join
        pg_roles r on a.grantee = r.oid
    where
        a.privilege_type = 'MAINTAIN'
) sub
order by schema_order, schema_name, privilege_type, grantee, default_for;
  schema_name   | privilege_type |      grantee       |     default_for     
----------------+----------------+--------------------+---------------------
 public         | MAINTAIN       | anon               | supabase_admin
 public         | MAINTAIN       | anon               | postgres
 public         | MAINTAIN       | authenticated      | supabase_admin
 public         | MAINTAIN       | authenticated      | postgres
 public         | MAINTAIN       | postgres           | supabase_admin
 public         | MAINTAIN       | postgres           | postgres
 public         | MAINTAIN       | service_role       | supabase_admin
 public         | MAINTAIN       | service_role       | postgres
 auth           | MAINTAIN       | dashboard_user     | supabase_auth_admin
 auth           | MAINTAIN       | postgres           | supabase_auth_admin
 extensions     | MAINTAIN       | postgres           | supabase_admin
 graphql        | MAINTAIN       | anon               | supabase_admin
 graphql        | MAINTAIN       | authenticated      | supabase_admin
 graphql        | MAINTAIN       | postgres           | supabase_admin
 graphql        | MAINTAIN       | service_role       | supabase_admin
 graphql_public | MAINTAIN       | anon               | supabase_admin
 graphql_public | MAINTAIN       | authenticated      | supabase_admin
 graphql_public | MAINTAIN       | postgres           | supabase_admin
 graphql_public | MAINTAIN       | service_role       | supabase_admin
 pgsodium       | MAINTAIN       | pgsodium_keyholder | supabase_admin
 pgsodium_masks | MAINTAIN       | pgsodium_keyiduser | supabase_admin
 realtime       | MAINTAIN       | dashboard_user     | supabase_admin
 realtime       | MAINTAIN       | postgres           | supabase_admin
 repack         | MAINTAIN       | postgres           | supabase_admin
 storage        | MAINTAIN       | anon               | postgres
 storage        | MAINTAIN       | authenticated      | postgres
 storage        | MAINTAIN       | postgres           | postgres
 storage        | MAINTAIN       | service_role       | postgres
(28 rows)

